pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/AbstractRule.java

```diff
@@ -378,8 +381,32 @@ public boolean equals(Object o) {
 
         if (equality) {
             Rule that = (Rule) o;
-            equality = getName().equals(that.getName()) && getPriority().equals(that.getPriority())
-                    && getPropertiesByPropertyDescriptor().equals(that.getPropertiesByPropertyDescriptor());
+            equality = getName().equals(that.getName()) && getPriority().equals(that.getPriority());
+
+            if (equality) {
+                // Before calling equals on the properties, replace any Pattern with it pattern string
+                // This is only needed for RegexProperties.
+                Map<PropertyDescriptor<?>, Object> propertiesWithValues = new HashMap<>();
+                getPropertiesByPropertyDescriptor()
+                        .forEach((key, value) -> {
+                            Object newValue = value;
+                            if (newValue instanceof Pattern) {
+                                newValue = ((Pattern) newValue).pattern();
+                            }
+                            propertiesWithValues.put(key, newValue);
+                        });
+                Map<PropertyDescriptor<?>, Object> thatPropertiesWithValues = new HashMap<>();
+                that.getPropertiesByPropertyDescriptor()
+                        .forEach((key, value) -> {
+                            Object newValue = value;
+                            if (newValue instanceof Pattern) {
+                                newValue = ((Pattern) newValue).pattern();
+                            }
+                            thatPropertiesWithValues.put(key, newValue);
+                        });
+
+                equality = propertiesWithValues.equals(thatPropertiesWithValues);
```

Comment:
First of all, amazing catch!

I'd consider an alternative approach…

1. Have `PropertySource` expose a method for property equality.
2. Have an implementation of that in `AbstractPropertySource` that instead of relying on `Map.equals`, checks wether the keysets match, and delegates to the `PropertyDescriptor` the equality check between 2 values for that property.

I think that is going to be clearer, applies to all `PropertySource`s, and forces any future `PropertyDescriptor` to provide the equality check without us having to remember to come and change this.

Reply:
I like your approach and agree that the problem is mostly that we had been using Map.equals. I think your new approach should be considered when implementing #3869. Instead of handling Maps of properties, we should be using `PropertyBundle` as a strongly typed container for properties and their values. Then we can just use `rule.getProperties().equals(otherRule.getProperties())` to implement equals, where `getProperties()` returns a property bundle, and equals is implemented like you describe.

I would prefer we reconsider that after #4380 is merged.

Reply:
I've moved the implementation of equals now into AbstractPropertySource. Added some more tests for some edge cases (property descriptors, defaults, values).

I hope, this change doesn't make it more difficult to move the equals whereever it goes with #4380 (the comment there in PropertyDescriptor says, it should go to Rule::equals).
Btw. we only compare currently the name of a property descriptor and neither description nor default value. #4380 changes that to include the description.